kernel/src/hil/i2c.rs

```diff
@@ -55,9 +14,25 @@ pub trait I2CMaster {
     fn set_master_client(&self, master_client: &'static dyn I2CHwMasterClient);
     fn enable(&self);
     fn disable(&self);
-    fn write_read(&self, addr: u8, data: &'static mut [u8], write_len: u8, read_len: u8);
-    fn write(&self, addr: u8, data: &'static mut [u8], len: u8);
-    fn read(&self, addr: u8, buffer: &'static mut [u8], len: u8);
+    fn write_read(
+        &self,
+        addr: u8,
+        data: &'static mut [u8],
+        write_len: u8,
+        read_len: u8,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])>;
```

Comment:
I thought the `i2c::Error` type was actually very useful, because it allowed you to easily figure out what had gone wrong, e.g., when using new hardware. I'm mostly in favor of keeping it, but enlarging it to include other general error conditions to support virtualization (e.g., OFF).

If the trait does transition to `ErrorCode`, then I think the calls need to map I2C error conditions to `ErrorCodes` and document that mapping. E.g., `ErrorCode::NOACK` could be mean either address NAK or data NAK, `ErrorCode::RESERVE` could mean arbitration lost, etc.

Reply:
(Sorry I am just getting to this now)

It seems like the issue we're wrestling with is neither the original Error nor ErrorCode can capture all of the error cases we care about, and we don't want to lose valuable information due to this. I'd suggest expanding Error to include relevant more general ErrorCodes, and writing a `from` so you can map it to an ErrorCode, with I2C-specific errors mapping as best we can.

Reply:
Indeed, that sounds like a very reasonable approach. I agree the custom type adds some value here, and the fact that we're going to have to convert it to an application specific `ErrorCode` down the line is just a consequence of Rust's approach to error handling, but one which is necessary in order to get correct and dependable interfaces.

I think this discussion has brought a lot of valuable insights on how we want to do error handling in these potentially more complex scenarios. I hope @alexandruradovici won't mind changing the error handling story again.

Reply:
Don worry, I don't mind, this is why I opened the discussion to get this discussion going. Just to sum up, to see if I got this right:
 - put back `i2c:Error` and make the HIL return this
 - add a From trait implementation to `ErrorCode` to be able to convert `i2c:Error` to `ErrorCode`

Another item is the is not the mapping of `i2c::Error` to `ErrorCode`. Based on the discussion here, this might be the following:

| i2c::Error | ErrorCode |
|----------|------------|
| DataNak | NOACK |
| AddressNak | NOACK |
| ArbitrationLost | REVERVE |
| Overrun | ? |
| Not Supported | NOSUPPORT |

Reply:
I might suggest `SIZE` for Overrun, though it is not a perfect match

Reply:
I agree it is not a perfect match, but as long as we clearly state the mapping, that is OK.

Reply:
Right. And IMHO it wouldn't even hurt if we didn't have an `impl From<i2c::Error> for ErrorCode`. Doing the conversion at the client ensures that the errors are mapped exactly according to its desired interface. But having a conversion to `ErrorCode` using the `From` trait is of course also fine, it's good to me either way!

Reply:
Another quick question: what error should be returned if the device has another operation inflight (the former `ErrorCode::BUSY`)? I was thinking of `Error::ArbitrationLost` but I am not sure if this is the best one.

Maybe we should add another item to `i2c::Error` for this case?